### PR TITLE
Lower MaxPooling and AveragePool to Krnl dialect using AffineMap

### PR DESCRIPTION
**Pull request by tungld**
This patch is to implement common functions for lowering pooling ops where:
- It supports padding directly in the lowering, thus there is no need to pad the input separately.
    - Remove this rewriting rule: `MaxPoolSingleOut(X, pads) = MaxPoolSinglePut(onnx.PadConstantValuePadOp(X,pads), pads = [0, 0, 0, 0])`
- It uses AffineMap to compute a pooling window for border edges instead of using a fixed pooling window of kernel size.
    - Helper function `pushAffineBound(AffineMap map, ArrayRef<Value> operands)` is newly added to help add an AffineMap as a bound for KrnlIterateOp.

TODO:
- [x] Add MLIR tests

### CODE REVIEW

File: src/Transform/ONNX/ONNXRewrite.td [hunk 1]
```diff
@@ -93,4 +102,39 @@ def MaxPoolSingleOutOpPaddingPattern: Pat<
...
+//
+// as:
+// %0 = onnx.PadConstantValuePadOp(%D)
+//     {pads = [0, 0, b0, b1, ... bK, 0, 0, e0, e1, ..., eK]} ->
+//     tensor<DPaddedShape>
+// %1 = onnx.AveragePool(%0 : tensor<DPaddedShape>) {pads = [0, ..., 0]} ->
+//     tensor<OutShape>
+//===----------------------------------------------------------------------===//
+
+// Pad values are NaNs if count_include_pad is disabled (default).
```
Comment (AlexandreEichenberger) [inline src/Transform/ONNX/ONNXRewrite.td]: I have a high level question: if you do an average with count_include_pad set to false. Do you adjust the denominator?

e.g. avg(4,6, pad, pad) = (4+6)/2 5?

Comment (tungld) [reply src/Transform/ONNX/ONNXRewrite.td]: Yes, I just adjust the denominator. NaN is signal to know a value is pad or not. By default `count_include_pad` is false. Please look at e2e tests in `test/backend/test.py` where passed e2e tests for AveragePool were added. So far all e2e tests for AveragePool passed.

File: src/Conversion/ONNXToKrnl/NN/Pooling.cpp [hunk 2]
```diff
@@ -177,141 +329,197 @@ struct ONNXMaxPoolSingleOutOpLowering : public ConversionPattern {
     //         R[n][c][r1][r2] = negative_infinity;
     //         for k1 = 0 .. KH:
     //           for k2 = 0 .. KW:
-    //             t = D[n][c][s1 * r1 + k1 * d1][s2 * r2 + k2 * d2];
+    //             t = D[n][c][s1 * r1 + k1][s2 * r2 + k2];
```
Comment (tjingrant) [inline src/Conversion/ONNXToKrnl/NN/Pooling.cpp]: Please update the pesudo code to reflect the algorithm of average pooling. Having a separate section explaining AveragePooling's algorithm may be desirable. I wouldn't worry about duplication of documentation.

File: src/Conversion/ONNXToKrnl/NN/Pooling.cpp [hunk 3]
```diff
@@ -31,44 +43,253 @@ Value mapToLowerScalarOp<ONNXMaxPoolSingleOutOp>(Operation *op,
...
+template <>
+bool getCountIncludePad<ONNXAveragePoolOp>(ONNXAveragePoolOp poolOp) {
+  return (poolOp.count_include_pad() == 1);
+}
+
+//===----------------------------------------------------------------------===//
+// Helper function to do post-processing after applying a filter window.
+//
+template <typename PoolOp>
+void doPostProcessingForPooling(ConversionPatternRewriter &rewriter,
```
[the remark this answers is not in the document]
Comment (tungld) [reply src/Conversion/ONNXToKrnl/NN/Pooling.cpp]: It will be also used for LpPool in the future where we need to compute p-norm.

File: src/Conversion/ONNXToKrnl/NN/Pooling.cpp [hunk 4]
```diff
@@ -31,44 +43,253 @@ Value mapToLowerScalarOp<ONNXMaxPoolSingleOutOp>(Operation *op,
...
+  return false;
+}
+
+// AveragePool
+template <>
+bool getCountIncludePad<ONNXAveragePoolOp>(ONNXAveragePoolOp poolOp) {
+  return (poolOp.count_include_pad() == 1);
+}
+
+//===----------------------------------------------------------------------===//
```
[the remark this answers is not in the document]
Comment (tungld) [reply src/Conversion/ONNXToKrnl/NN/Pooling.cpp]: I see the same format in ONNXOps.cpp where shape inference is implemented. Could you teach me what convention we are using? Thanks!

File: src/Conversion/ONNXToKrnl/NN/Pooling.cpp [hunk 5]
```diff
@@ -31,44 +43,253 @@ Value mapToLowerScalarOp<ONNXMaxPoolSingleOutOp>(Operation *op,
...
+    zero = rewriter.create<ConstantOp>(
+        loc, rewriter.getIntegerAttr(rewriter.getIntegerType(64), 0));
+  }
+  one = rewriter.create<ConstantOp>(
+      loc, rewriter.getIntegerAttr(rewriter.getIntegerType(64), 1));
+
+  int spatialRank = resultRank - kernelOffset;
+  int64_t dilation = 1;
+  for (int i = kernelOffset; i < resultShape.size(); ++i) {
+    if (resultShape[i] < 0) {
```
[the remark this answers is not in the document]
Comment (tungld) [reply src/Conversion/ONNXToKrnl/NN/Pooling.cpp]: Yes, it's great to have such a function.

**Comment by tjingrant** [comment]
Hi @tungld , I wrote a simple script to illustrate the code for computing average pooling when `count_include_pad` is false without explicitly counting/calculating the number of pooled pixels. I verified it against pytorch for `ceil_mode = True`, but I cannot verify the case where dilation is non-trivial because PyTorch average pool does not support dilation, but I think it's probably not too far from being correct either. The point I'm trying to make is that we can use some closed-form expression to compute the divisor used in the averaging process, which will better our performance. 

Anyways, here's the python script. Feel free to ping me for a discussion Slack, we are in similar timezones these days.
```python
import torch
import torch.nn.functional as F
import numpy as np
import random

from math import floor, ceil

def cmp_avg_pool(N, C, H, W, kH, kW, sH, sW, dH, dW, ceil_mode):
	my_round = ceil if ceil_mode else floor

	imgs = torch.rand([N, C, W, W])
	pooled = F.avg_pool2d(imgs,
	                      kernel_size=[kH, kW],
	                      stride=[sH, sW],
	                      padding=0,
	                      ceil_mode=ceil_mode,
	                      count_include_pad=False,
	                      divisor_override=None)
	imgs, pooled = imgs.numpy(), np.nan_to_num(pooled.numpy())

	# Compute output shape:
	H_out = my_round(float(H - kH) / float(sH) + 1.0)
	W_out = my_round(float(W - kW) / float(sW) + 1.0)
	my_pooled = np.zeros([
	    N, C,
	    H_out, W_out
	])
	# Make sure shape is as expected.
	assert my_pooled.shape == pooled.shape


	for n in range(N):
		for c in range(C):
			for ho in range(H_out):
				for wo in range(W_out):
					for hi in range(ho * sH, min(H, ho * sH + kH * dH), dH):
						for wi in range(wo * sW, min(W, wo * sW + kW * dW), dW):
							my_pooled[n, c, ho, wo] += imgs[n, c, hi, wi]
					# Compute the divisor for average pooling:
					h_count = my_round(float(min(H, ho * sH + kH * dH) - ho * sH) / float(dH))
					w_count = my_round(float(min(W, wo * sW + kW * dW) - wo * sW) / float(dW))
					if h_count * w_count > 0:
						my_pooled[n, c, ho, wo] = my_pooled[n, c, ho, wo] / (h_count * w_count)

	np.testing.assert_array_almost_equal(my_pooled, pooled)

def main():
	N = random.randint(1, 10)
	C = random.randint(1, 10)
	H = random.randint(20, 40)
	W = H

	kH = random.randint(1, 7)
	kW = kH

	sH = random.randint(1, 5)
	sW = sH

	dH = 1
	dW = 1

	ceil_mode = random.choice([True, False])

	cmp_avg_pool(N, C, H, W, kH, kW, sH, sW, dH, dW, ceil_mode)

for i in range(100):
	main()
```

**Comment by tungld** [comment]
@tjingrant Could you please quickly review changes related to use of AffineMap? I added some helper functions in `KrnlHelper` that push AffineMap as the upper bound for KrnlIterateOp. Please look at the commit `Push AffineMap as upper bound for KrnlIterateOp`

I am moving to AffineMap and the code is a bit messy now, so I will change this patch's status to WIP.

**Comment by tjingrant** [comment]
@tungld I think the changes are great; some comments may be helpful for others looking into the codebase, as the case where upper/lower bound is specified using AffineMap requires some learning/explaination.

**Comment by tungld** [comment]
@tjingrant @AlexandreEichenberger New implementation is ready to review. With the new implementation, we don't need to pad the input in advance, which would lead to better performance.

File: src/Conversion/ONNXToKrnl/NN/Pooling.cpp [hunk 6]
```diff
@@ -30,288 +44,519 @@ Value emitScalarOpFor<ONNXMaxPoolSingleOutOp>(
...
+      // the border edges. Thus, we will compute the start and end indices for
+      // each dimension as follows.
+      //   firstValidH = ceil(float(ptH / dH)) * dH - ptH
+      //   startH = max(firstValidH, ho * sH - ptH)
+      //   endH = min(H, ho * sH + (kH - 1) * dH  + 1 - pbH)
+      //   hDim = round(float(endH - startH) / float(dH))
+
+      // Prepare induction variables and constants as arguments for the affine
+      // maps.
+      SmallVector<SmallVector<Value, 4>, 4> IVsAndConstants;
```
Comment (tjingrant) [inline src/Conversion/ONNXToKrnl/NN/Pooling.cpp:392]: I would really appreciate if IVsAndConstants can be represented using nested structs to make the semantic meaning of both data structure obvious for readers.
